File: ecg-telemetry-project/ios/ECGMonitor/DSP/BeatClassifier.swift
import Foundation

/// Beat classifier — Swift port of `RuleClassifier` / `LinearClassifier` from
/// `server/ecg_algorithms.py`.
///
/// Two modes:
///  • rules   — the textbook criteria; works with no training data at all.
///  • trained — multinomial logistic regression whose weights were fitted on
///    MIT-BIH by `server/train_classifier.py` and bundled as `model.json`.
///    Exactly the same weights the server uses.
///
/// Features are RATIOS against the patient's own running normal beat rather than
/// absolute millivolts and milliseconds, so one set of thresholds works across
/// leads, electrode placements and patients — absolute limits do not.
final class BeatClassifier {

    static let labels = ["Normal", "Supraventricular", "Ventricular", "Other"]

    private var mean: [Double] = []
    private var scale: [Double] = []
    private var coef: [[Double]] = []
    private var intercept: [Double] = []
    private(set) var isTrained = false

    init() {}

    /// Load weights exported by `train_classifier.py`.
    /// Returns nil if the file is missing or malformed, so the caller can fall
    /// back to the rules rather than shipping a half-initialised model.
    init?(modelJSON data: Data) {
        guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let m = obj["mean"] as? [Double],
              let s = obj["scale"] as? [Double],
              let c = obj["coef"] as? [[Double]],
              let b = obj["intercept"] as? [Double]
        else { return nil }
        mean = m
        scale = s
        coef = c
        intercept = b
        isTrained = true
    }

    /// Convenience loader for a bundled `model.json`.
    static func fromBundle() -> BeatClassifier {
        if let url = Bundle.main.url(forResource: "model", withExtension: "json"),
           let data = try? Data(contentsOf: url),
           let clf = BeatClassifier(modelJSON: data) {
            return clf
        }
        return BeatClassifier()
    }

    /// Feature order must match `FEATURE_NAMES` in `ecg_algorithms.py`.
    static func features(_ b: PanTompkins.Beat) -> [Double] {
        let m = b.rrMean > 0 ? b.rrMean : 0.8
        let wm = b.widthMean != 0 ? b.widthMean : 1e-6
        let am = b.ampMean != 0 ? b.ampMean : 1e-6
        return [
            b.rrPrev,
            b.rrNext,
            b.rrPrev / m,
            b.rrNext / m,
            b.width / wm,
            b.amplitude / am,
            b.corr,
            (b.rrNext - b.rrPrev) / m,
            b.width,
        ]
    }

    /// Sets `label` and `confidence` in place.
    func classify(_ b: inout PanTompkins.Beat) {
        let x = BeatClassifier.features(b)
        if isTrained { predictLinear(&b, x) } else { predictRules(&b, x) }
    }

    private func predictLinear(_ b: inout PanTompkins.Beat, _ x: [Double]) {
        var z = [Double](repeating: 0, count: x.count)
        for i in 0..<x.count {
            let s = (i < scale.count && scale[i] != 0) ? scale[i] : 1.0
            z[i] = (x[i] - (i < mean.count ? mean[i] : 0)) / s
        }
        var scores = [Double](repeating: 0, count: coef.count)
        for k in 0..<coef.count {
            var s = intercept[k]
            for i in 0..<min(z.count, coef[k].count) { s += coef[k][i] * z[i] }
            scores[k] = s
        }
        let mx = scores.max() ?? 0
        var tot = 0.0
        for k in 0..<scores.count {
            scores[k] = exp(scores[k] - mx)
            tot += scores[k]
        }
        var best = 0
        for k in 1..<scores.count where scores[k] > scores[best] { best = k }
        b.label = BeatClassifier.labels[min(best, BeatClassifier.labels.count - 1)]
        b.confidence = scores[best] / tot
    }

    private func predictRules(_ b: inout PanTompkins.Beat, _ x: [Double]) {
        let rrRatio = x[2]
        let rrNextRatio = x[3]
        let widthRatio = x[4]
        let ampRatio = x[5]
        let corr = x[6]

        if ampRatio < 0.25 || widthRatio > 3.5 {
            b.label = "Other"; b.confidence = 0.6; return      // noise / lead-off
        }

        let premature = rrRatio < 0.85
        let compensatory = rrNextRatio > 1.10

        // The bedside discriminator between a PVC and an atrial premature beat is
        // the pause that follows. A PVC does not reset the sinus node, so the
        // next sinus beat arrives on schedule and the pause is fully
        // compensatory; an atrial ectopic resets it and the pause is shorter.
        let fullPause = rrNextRatio > 1.20

        // Ventricular origin means an abnormal activation path, showing up as a
        // complex that is both unlike the patient's normal AND broader. Demanding
        // both keeps ordinary beats next to an ectopic — whose correlation window
        // is contaminated by the neighbour — out of the class.
        let ventricular = (corr < 0.80 && (widthRatio > 1.05 || fullPause))
                          || widthRatio > 1.30

        if ventricular {
            b.label = "Ventricular"
            b.confidence = (premature && compensatory) ? 0.9 : 0.7
        } else if premature {
            b.label = "Supraventricular"
            b.confidence = compensatory ? 0.8 : 0.65
        } else if rrRatio > 1.6 {
            b.label = "Other"
            b.confidence = 0.5                                  // dropped / escape
        } else {
            b.label = "Normal"
            b.confidence = 0.9
        }
    }
}

File: android/app/src/main/java/com/example/ecg/dsp/BeatClassifier.java
package com.example.ecg.dsp;

import org.json.JSONArray;
import org.json.JSONObject;

/**
 * Beat classifier - the port of RuleClassifier / LinearClassifier from
 * server/ecg_algorithms.py.
 *
 * Two modes:
 *   - RULES: encodes the textbook criteria, works with no training data.
 *   - TRAINED: multinomial logistic regression whose weights were fitted on the
 *     MIT-BIH Arrhythmia Database by server/train_classifier.py and dropped into
 *     app/src/main/assets/model.json. Exactly the same weights the server uses.
 *
 * Features are RATIOS against the patient's own running normal beat rather than
 * absolute millivolts and milliseconds. That way one set of thresholds works
 * across leads, electrode placements and patients, which absolute limits do not.
 */
public class BeatClassifier {

    public static final String[] LABELS =
        {"Normal", "Supraventricular", "Ventricular", "Other"};

    private double[] mean, scale, intercept;
    private double[][] coef;
    private boolean trained = false;

    // Decision tree, as written by train_classifier.py. Arrays are parallel and
    // indexed by node; feature[n] < 0 marks a leaf.
    private int[] tFeature, tLeft, tRight;
    private double[] tThreshold;
    private double[][] tValue;
    private boolean isTree = false;

    public BeatClassifier() { }

    /** Load weights exported by train_classifier.py. Falls back to rules on error. */
    public BeatClassifier(String modelJson) {
        try {
            JSONObject o = new JSONObject(modelJson);
            if ("tree".equals(o.optString("type", "linear"))) {
                tFeature = toIntArray(o.getJSONArray("feature"));
                tLeft = toIntArray(o.getJSONArray("left"));
                tRight = toIntArray(o.getJSONArray("right"));
                tThreshold = toArray(o.getJSONArray("threshold"));
                JSONArray v = o.getJSONArray("value");
                tValue = new double[v.length()][];
                for (int i = 0; i < v.length(); i++) tValue[i] = toArray(v.getJSONArray(i));
                isTree = true;
                trained = true;
                return;
            }
            mean = toArray(o.getJSONArray("mean"));
            scale = toArray(o.getJSONArray("scale"));
            intercept = toArray(o.getJSONArray("intercept"));
            JSONArray c = o.getJSONArray("coef");
            coef = new double[c.length()][];
            for (int i = 0; i < c.length(); i++) coef[i] = toArray(c.getJSONArray(i));
            trained = true;
        } catch (Exception e) {
            trained = false;   // model.json missing or malformed -> use rules
        }
    }

    public boolean isTrained() { return trained; }

    private static int[] toIntArray(JSONArray a) {
        int[] r = new int[a.length()];
        for (int i = 0; i < a.length(); i++) r[i] = a.optInt(i, -1);
        return r;
    }

    private static double[] toArray(JSONArray a) {
        double[] r = new double[a.length()];
        for (int i = 0; i < a.length(); i++) r[i] = a.optDouble(i, 0);
        return r;
    }

    /** Feature order must match FEATURE_NAMES in ecg_algorithms.py. */
    public static double[] features(PanTompkins.Beat b) {
        double m = b.rrMean > 0 ? b.rrMean : 0.8;
        double wm = b.widthMean != 0 ? b.widthMean : 1e-6;
        double am = b.ampMean != 0 ? b.ampMean : 1e-6;
        return new double[] {
            b.rrPrev,
            b.rrNext,
            b.rrPrev / m,
            b.rrNext / m,
            b.width / wm,
            b.amplitude / am,
            b.corr,
            (b.rrNext - b.rrPrev) / m,
            b.width,
            b.zcr / (b.zcrMean != 0 ? b.zcrMean : 1e-6),
            b.area / (b.areaMean != 0 ? b.areaMean : 1e-6),
        };
    }

    /** Classifies in place: sets beat.label and beat.confidence. */
    public void classify(PanTompkins.Beat b) {
        double[] x = features(b);
        if (isTree) predictTree(b, x);
        else if (trained) predictLinear(b, x);
        else predictRules(b, x);
    }

    /** Walk the tree: left when the feature is <= threshold, matching
     *  scikit-learn's convention exactly. */
    private void predictTree(PanTompkins.Beat b, double[] x) {
        int node = 0;
        while (tFeature[node] >= 0) {
            node = (x[tFeature[node]] <= tThreshold[node]) ? tLeft[node] : tRight[node];
        }
        double[] probs = tValue[node];
        int best = 0;
        for (int k = 1; k < probs.length; k++) if (probs[k] > probs[best]) best = k;
        b.label = LABELS[Math.min(best, LABELS.length - 1)];
        b.confidence = probs[best];
    }

    private void predictLinear(PanTompkins.Beat b, double[] x) {
        double[] z = new double[x.length];
        for (int i = 0; i < x.length; i++) {
            double s = (i < scale.length && scale[i] != 0) ? scale[i] : 1.0;
            z[i] = (x[i] - (i < mean.length ? mean[i] : 0)) / s;
        }
        double[] scores = new double[coef.length];
        double mx = -Double.MAX_VALUE;
        for (int k = 0; k < coef.length; k++) {
            double s = intercept[k];
            for (int i = 0; i < z.length && i < coef[k].length; i++) s += coef[k][i] * z[i];
            scores[k] = s;
            if (s > mx) mx = s;
        }
        double tot = 0;
        for (int k = 0; k < scores.length; k++) { scores[k] = Math.exp(scores[k] - mx); tot += scores[k]; }
        int best = 0;
        for (int k = 1; k < scores.length; k++) if (scores[k] > scores[best]) best = k;
        b.label = LABELS[Math.min(best, LABELS.length - 1)];
        b.confidence = scores[best] / tot;
    }

    private void predictRules(PanTompkins.Beat b, double[] x) {
        double rrRatio = x[2], rrNextRatio = x[3];
        double widthRatio = x[4], ampRatio = x[5], corr = x[6];
        double zcrRatio = x[9], areaRatio = x[10];

        if (ampRatio < 0.25 || widthRatio > 3.5) { set(b, "Other", 0.6); return; }

        boolean premature = rrRatio < 0.85;
        boolean compensatory = rrNextRatio > 1.10;

        // The bedside discriminator between a PVC and an atrial premature beat
        // is the pause that follows. A PVC does not reset the sinus node, so
        // the next sinus beat arrives on schedule and the pause is fully
        // compensatory; an atrial ectopic resets it and the pause is shorter.
        boolean fullPause = rrNextRatio > 1.20;

        // Abnormal morphology is NECESSARY, not merely one vote. A beat that
        // still matches the patient's own template is by definition being
        // conducted the normal way, whatever its width measurement says. Width
        // alone was tried as a sufficient condition and flagged large numbers of
        // ordinary beats whose measurement window was contaminated by a
        // neighbouring complex.
        boolean different = corr < 0.80;

        // Shape difference is DIRECTION-AGNOSTIC. Measured in the 5-15 Hz
        // detection band a ventricular complex comes out NARROWER than a sinus
        // one, not wider: the bandpass strips the low-frequency content that
        // makes a PVC broad, collapsing it into one smooth concentrated hump,
        // while a sinus QRS keeps its sharp biphasic R-S spread. On MIT-BIH
        // record 233 annotated PVCs measured widthRatio 0.39-0.74 against a
        // normal of 1.0. An earlier version tested only for widening and so had
        // the inequality backwards.
        boolean reshaped = !(widthRatio >= 0.85 && widthRatio <= 1.12);
        boolean requantified = !(areaRatio >= 0.83 && areaRatio <= 1.20);
        boolean slow = zcrRatio < 0.80 || zcrRatio > 1.25;
        boolean ventricular = different && (reshaped || requantified || slow);

        if (ventricular) { set(b, "Ventricular", (premature && compensatory) ? 0.9 : 0.7); return; }
        if (premature)   { set(b, "Supraventricular", compensatory ? 0.8 : 0.65); return; }
        if (rrRatio > 1.6) { set(b, "Other", 0.5); return; }   // dropped / escape beat
        set(b, "Normal", 0.9);
    }

    private void set(PanTompkins.Beat b, String label, double conf) {
        b.label = label;
        b.confidence = conf;
    }
}
